### PR TITLE
[perf experiments] Clone all MIR bodies

### PR DESCRIPTION
**Pull request by saethlin**
We keep saying things like cloning all the MIR would be expensive, but... how expensive actually?

Based on this PR, it looks like the literal clone is cheap enough to be considered free, but running basically any MIR transform on the cloned MIR is a measurable slowdown. The _hope_ here was that we could do optimizations on monomorphized MIR. I think ther eare two problems with that:

The first problem is this: https://github.com/rust-lang/rust/blob/032be6f7bbe091c7dfa29f115e94b9cc9bae1758/compiler/rustc_mir_transform/src/add_call_guards.rs#L28

LLVM needs a certain structure around calls, which is generally wasteful in terms of the MIR that it takes to generate. We currently cope with this by creating the required structure at the end of the `optimized_mir` passes. This is very problematic, because the most obvious post-mono optimization is running something like `SimplifyCfg` to remove `if true` and `if false` and the resultant goto chains, but that deletes all the work of `AddCallGuard`, so now you need to run _at least_ two post-mono MIR transforms.

The second problem is that naively monomorphizing the entire Body doesn't skip unreachable blocks. I tried adding some code that uses the same reachability analysis to skip dead code, but something subtle happens that causes one of our secondary benchmarks to regress a lot. That's this perf run: https://github.com/rust-lang/rust/pull/125025#issuecomment-2156198087. I never figured out what causes this slowdown. The later perf runs are me trying to figure out what that's from.

### CODE REVIEW

**Comment by saethlin** [comment]
Oops, sorry about that.

@bors try @rust-timer queue

File: compiler/rustc_codegen_ssa/src/mir/mod.rs [hunk 1]
```diff
@@ -166,6 +166,7 @@ pub fn codegen_mir<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
     let llfn = cx.get_fn(instance);
 
     let mir = cx.tcx().instance_mir(instance.def);
+    drop(std::hint::black_box(mir.clone()));
```
Comment (compiler-errors) [inline compiler/rustc_codegen_ssa/src/mir/mod.rs:169]: Are you certain that you're copying `mir::Body` and not `&mir::Body`?

Comment (saethlin) [reply compiler/rustc_codegen_ssa/src/mir/mod.rs:169]: I am not certain.

Comment (compiler-errors) [reply compiler/rustc_codegen_ssa/src/mir/mod.rs:169]: Actually, I'm pretty certain now. `&mir::Body` will match against `Clone::clone` without applying autoref first, which clones `mir::Body`.

Comment (saethlin) [reply compiler/rustc_codegen_ssa/src/mir/mod.rs]: Normally I'd just assume you know what you're talking about, but this result is quite remarkable if true, so I'm going to be doubly sure.

Comment (compiler-errors) [reply compiler/rustc_codegen_ssa/src/mir/mod.rs]: yep :+1: there's a chance there's something *else* weird happening here too.

Comment (Kobzol) [reply compiler/rustc_codegen_ssa/src/mir/mod.rs]: Calling `(&T).clone()` should call `clone()` on `T` (even though `&T` also implements `Clone`, which is true for all types).
I would be more worried about the black box. Maybe you could try cloning the `mir` variable into `mir2`, and then continue using `mir2` in the code? (or just shadow the original variable).

Comment (RalfJung) [reply compiler/rustc_codegen_ssa/src/mir/mod.rs]: `mir::Body::clone` should avoid the ambiguity.

**Comment by RalfJung** [comment]
Hm... so maybe Miri should clone the body and apply the substitution once, rather than applying it on each statement? For hot loops that will definitely help a lot. But for big functions where we only run a small part of the code, less so.

**Comment by saethlin** [comment]
> Hm... so maybe Miri should clone the body and apply the substitution once, rather than applying it on each statement?

Could be. The idea here is to evaluate whether we can do this same thing in codegen so that we can have post-mono optimization passes instead of the tangled mess we have right now where optimizations are woven into the fundamental process of lowering.

I don't know that monomorphization is particularly hot in the interpreter. I certainly don't remember poking around it in a profile.

**Comment by saethlin** [comment]
The only substantial regression is in a microbenchmark with unrealistic code. Onwards!

**Comment by saethlin** [comment]
I've updated the PR description with my conclusions from these experiments.

**Pull request closed without merging**